c/cert/src/rules/FIO32-C/DoNotPerformFileOperationsOnDevices.ql

```diff
@@ -45,20 +46,93 @@ class FileFunction extends FunctionWithWrappers {
   override predicate interestingArg(int arg) { arg = 0 }
 }
 
-class TaintedPathConfiguration extends TaintTrackingConfiguration {
-  override predicate isSink(Element tainted) {
-    exists(FileFunction fileFunction | fileFunction.outermostWrapperFunctionCall(tainted, _))
+Expr asSourceExpr(DataFlow::Node node) {
+  result = node.asConvertedExpr()
+  or
+  result = node.asDefiningArgument()
+}
+
+Expr asSinkExpr(DataFlow::Node node) {
+  result =
+    node.asOperand()
+        .(SideEffectOperand)
+        .getUse()
+        .(ReadSideEffectInstruction)
+        .getArgumentDef()
+        .getUnconvertedResultExpression()
+}
+
+/**
+ * Holds for a variable that has any kind of upper-bound check anywhere in the program.
+ * This is biased towards being inclusive and being a coarse overapproximation because
+ * there are a lot of valid ways of doing an upper bounds checks if we don't consider
+ * where it occurs, for example:
+ * ```cpp
+ *   if (x < 10) { sink(x); }
+ *
+ *   if (10 > y) { sink(y); }
+ *
+ *   if (z > 10) { z = 10; }
+ *   sink(z);
+ * ```
+ */
+predicate hasUpperBoundsCheck(Variable var) {
+  exists(RelationalOperation oper, VariableAccess access |
+    oper.getAnOperand() = access and
```

Comment:
Because we are looking at an upper bound check, we can use the `getLesserOperand` for the variable access.

```suggestion
    oper.getLesserOperand() = access and
```

Reply:
We could, but that'll probably need some investigation of how this affects alerts. The current version of this predicate is a direct copy of what `DefaultTaintTracking` does to ensure we do not change the result set too much.

Note that we expect to do further rewrites of this query, but we currently miss some models on the CodeQL side to make this possible.

Reply:
A proper rewrite of this query on the CodeQL-side will involve getting rid of this predicate. So, leaving this as-is for now.